🚀 3단계 - 기능 우선 패키지 구성하기

jivebreaddev opened the pull request:
시훈님 이번에도 잘 부탁드리겠습니다!

- domain 별 분리

포트어댑터 패턴으로 아키텍처 생성
![image](https://github.com/next-step/ddd-strategic-design/assets/89639413/300eb6bf-6422-414a-a2d9-963a5a8432fd)
![image](https://github.com/next-step/ddd-strategic-design/assets/89639413/cab35b37-fa69-4d62-8875-e58cb524f8f7)

![image](https://github.com/next-step/ddd-strategic-design/assets/89639413/b16c289b-8a9b-4f66-b216-e9c439f04aee)



top project
|---- common (순수 자바 패키지)
|----|---- exception (오류 패키지)
|--- application (SERVICE 로직)
|----|---- Service
|---- port 
|----|---- out
|----|---- |----- Repository
|----|---- in
|----|---- |----- ServiceInterface
|    domain (도메인 레이어)
|----|----- Order
|----|----- Menu
|----|----- Product
|----persistence  (영속성 레이어)
|----ui  (외부 노출 인터페이스 레이어)
|----|----- Controller
|----infra  (영속성 레이어 + 클라이언트 외부 의존성)
|----|----- Client
|----|----- JPA INTERFACES

Jivebread committed:
chore(packages restructuring): Package 설계

- domain 별 분리

포트어댑터 패턴으로 아키텍처 생성

top project
|--- common (순수 자바 패키지)
|    |---- exception (오류 패키지)
|--- application (SERVICE 로직)
|    |---- Service
|    port (영속성 인터페이스)
|    |     out
|    |     |----- Repository
|    |     in
|    |     |----- ServiceInterface
|    domain (도메인 레이어)
|    |----- Order
|    |----- Menu
|    |----- Product
|    persistence  (영속성 레이어)
|    ui  (외부 노출 인터페이스 레이어)
|    |----- Controller
|    infra  (영속성 레이어 + 클라이언트 외부 의존성)
|    |----- Client
|    |----- JPA INTERFACES

sihoony commented on src/main/java/kitchenpos/menu/application/MenuGroupServiceImpl.java:
```diff
@@ -1,7 +1,8 @@
-package kitchenpos.application;
+package kitchenpos.menu.application;
```
👍 👍

sihoony commented on src/main/java/kitchenpos/menu/application/PurgomalumClient.java:
```diff
@@ -1,4 +1,4 @@
-package kitchenpos.infra;
+package kitchenpos.menu.application;
```
여러 Context에 사용되는 기능이라면 어떻게 분리하는게 좋을까요? Menu에 위치하는게 좋을지, 다른 package로 분리할지 고민하면 좋을 것 같네요.

sihoony commented on src/main/java/kitchenpos/order/application/KitchenridersClient.java:
```diff
@@ -1,4 +1,4 @@
-package kitchenpos.infra;
+package kitchenpos.order.application;
```
KitchenridersClient은 Interface로 어느 위치에 속하는게 좋을지 고민해보면 좋을 것 같습니다!!

sihoony reviewed (changes requested):
안녕하세요~! 시창님,
이번 미션도 잘 진행해 주셨네요~! 👍 
간단한 코멘트를 남겼으니 확인 부탁드립니다!! 
즐거운 하루되세요~! 😄

Jivebread committed:
chore(Modeling): kitchenpos 모델링
- 공통 외부 패키지 external으로 분리해서 사용
- application/port/out 에서 도메인으로 Repository들을 이전 (피드백 강의의 모듈 디자인을 따라 연습해보려는 의도) - 동일해 보인다.

jivebreaddev commented:
- 요청사항
1. 여러 컨텍스트에 속한 기능이라면 다른 패키지 분리
2. client interface를 어디다 분리할지 고민 필요


![image](https://github.com/next-step/ddd-strategic-design/assets/89639413/c7cd1e01-2cd8-4085-b23a-5c331969765d)




## 질문들

1. context 하위에 도메인들을  전부 나눠줘야하나요?
  a. 식당내 식사
  b. 배달 식사
  c. 포장 식사

3. 말씀하셨던것처럼 공통으로 사용하는 외부패키지의 경우 external으로 분리를 했습니다.
   - 저는 실무에서 infra_xxx_client와 같이 module로 나눠서 사용하게 됩니다.
   - 보통은 패키지 레벨로 나누시나요? 혹시 그렇다면 external이 비대해지지 않을까요?
   

4. 패키지간 의존관계의 경우에는 현재 단계에서 신경쓰지 않는것이 맞죠?
   - `간접참조로 연결성 끊어내는 것`은 다음주 미션들에서 진행하면 되는것으로 보여서 진행하지 않았습니다.

5. KitchenRidersClient의 인터페이스는 service에서 호출되게 될텐데
    infra -> service에 의존하는 구조를 위해서 KitchenRidersClient 인터페이스를 service에 둬야하는게 아닐까요? 
     `제가 생각하는 문제 해결 방안`
        1. external/kitchen 에는 defaultRiderClientImpl 같은 구현체들 만두고 사용하는 곳에서 interface를 정의해줘야 될것같습니다.
           - 만약에 다른 패키지에서 해당 `공통 구현체`를 사용하지 않는다면 `인터페이스 분리법칙`에 따라 다른 인터페이스를 생성해서 구현체를 따로 만들어서 생성해야 되지 않을까? 라는 생각입니다.


    domain에서는 infra/JpaRepository로 Dependency Inversion을 해줬었던것으로 보이는데 무엇이 다를까요?
    
    
 ## DIP 를 통해 INFRA가 SERVICE으로 의존성을 가지게 되는 사진

![image](https://github.com/next-step/ddd-strategic-design/assets/89639413/17e717b2-d08d-4757-af30-d387d266b898)

Jivebread committed:
chore(Modeling): kitchenpos 모델링
- product 패키지의 repository application 에서 domain으로 수정

sihoony commented:
> * 요청사항
> 
> 1. 여러 컨텍스트에 속한 기능이라면 다른 패키지 분리
> 2. client interface를 어디다 분리할지 고민 필요
> 
> ![image](https://private-user-images.githubusercontent.com/89639413/333845682-c7cd1e01-2cd8-4085-b23a-5c331969765d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTY3MTk0NzksIm5iZiI6MTcxNjcxOTE3OSwicGF0aCI6Ii84OTYzOTQxMy8zMzM4NDU2ODItYzdjZDFlMDEtMmNkOC00MDg1LWIyM2EtNWMzMzE5Njk3NjVkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTI2VDEwMjYxOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdkMjc0OTUyOWU3NzYyY2FjM2QxODYxZWNiYWRmZmUwMThkMWJlMjMyNWQ2YTRjZjU3NDU4Nzc3MjU1YWMxNWYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Q3-qqOk0MdLhr16xvJJ6KtR3pl5KHR2BY9HJCJOkSag)
> 
> ## 질문들
> 1. context 하위에 도메인들을  전부 나눠줘야하나요?
>    a. 식당내 식사
>    b. 배달 식사
>    c. 포장 식사
> 2. 말씀하셨던것처럼 공통으로 사용하는 외부패키지의 경우 external으로 분리를 했습니다.
>    
>    * 저는 실무에서 infra_xxx_client와 같이 module로 나눠서 사용하게 됩니다.
>    * 보통은 패키지 레벨로 나누시나요? 혹시 그렇다면 external이 비대해지지 않을까요?
> 3. 패키지간 의존관계의 경우에는 현재 단계에서 신경쓰지 않는것이 맞죠?
>    
>    * `간접참조로 연결성 끊어내는 것`은 다음주 미션들에서 진행하면 되는것으로 보여서 진행하지 않았습니다.
> 4. KitchenRidersClient의 인터페이스는 service에서 호출되게 될텐데
>    infra -> service에 의존하는 구조를 위해서 KitchenRidersClient 인터페이스를 service에 둬야하는게 아닐까요?
>    `제가 생각하는 문제 해결 방안`
>    1. external/kitchen 에는 defaultRiderClientImpl 같은 구현체들 만두고 사용하는 곳에서 interface를 정의해줘야 될것같습니다.
>    - 만약에 다른 패키지에서 해당 `공통 구현체`를 사용하지 않는다면 `인터페이스 분리법칙`에 따라 다른 인터페이스를 생성해서 구현체를 따로 만들어서 생성해야 되지 않을까? 라는 생각입니다.
>    domain에서는 infra/JpaRepository로 Dependency Inversion을 해줬었던것으로 보이는데 무엇이 다를까요?
> 
> ## DIP 를 통해 INFRA가 SERVICE으로 의존성을 가지게 되는 사진
> ![image](https://private-user-images.githubusercontent.com/89639413/333842558-17e717b2-d08d-4757-af30-d387d266b898.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTY3MTk0NzksIm5iZiI6MTcxNjcxOTE3OSwicGF0aCI6Ii84OTYzOTQxMy8zMzM4NDI1NTgtMTdlNzE3YjItZDA4ZC00NzU3LWFmMzAtZDM4N2QyNjZiODk4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTI2VDEwMjYxOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTAyOGYzNmIwZmMwYzRhMTIzNTBmMzI3ZjRhZWY0NWQ0YzNjOTk1ZmZlNzdhNDExNTk0YjIzYjVjZGRjZTdiNWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.EOVW4m4X2P6iHlx5rxrtbAf6patx1FiAnwBZxvLNPIY)

1. 식사를 하는 방식에 따른 유형으로 생각됩니다. 저는 굳이 나누진 않을 것 같네요. 이유로는 동일한 행위지만 유형에 따라 비즈니스 로직이 다르다고 보입니다.

2. 공통 사용 모듈은 공통으로 분리하여 사용합니다. 시창님이 이야기 하신것 처럼 비대해 질수 있어, 공통 모듈에 1Depth를 추가하여 어떤 모듈인지 특정하여 사용합니다. 다만, External 모듈이지만 특정 도메인에 속해 있다면 공토 모듈이 아닌, 사용 도메인에 위치하비니다.

3. 현재는 기존 Class에 대한 패키지 구분(Context 구분)으로 보시면 됩니다.

4. Trade-off라고 생각합니다. 다만, 도메인별 Interface가 존재한다면 영향도에 따라 모든 도메인을 수정하거나, 도메인별  version을 따로 가져가야할 수도 있을 것 같네요. 추가로,  Repo interface가 Domain이 아닌 다른 영역에 있다면 Domain이 다른 영역을 인지해야하는 상황잉 올 것 같네요.

sihoony reviewed (approved):
안녕하세요~! 시창님,
피드백 잘 반영해 주셨네요~! 👍 
그동안 고생하셨습니다. 즐거운 주말되세요~! 😄
